lib/common/rules.c

```diff
@@ -1004,89 +1004,116 @@ value_from_source(const char *value, enum pcmk__reference_source source,
  * \internal
  * \brief Evaluate a node attribute rule expression
  *
- * \param[in] expr        XML of a rule's PCMK_XE_EXPRESSION subelement
+ * \param[in] expression  XML of a rule's PCMK_XE_EXPRESSION subelement
  * \param[in] rule_data   Values used to evaluate rule criteria
  *
- * \return Standard Pacemaker return code
+ * \return Standard Pacemaker return code (\c pcmk_rc_ok if the expression
+ *         passes, some other value if it does not)
  */
 int
-pcmk__evaluate_attr_expression(const xmlNode *expr,
+pcmk__evaluate_attr_expression(const xmlNode *expression,
                                const pcmk_rule_data_t *rule_data)
 {
-    gboolean attr_allocated = FALSE;
-    const char *h_val = NULL;
-
-    const char *id = pcmk__xe_id(expr);
-    const char *attr = crm_element_value(expr, PCMK_XA_ATTRIBUTE);
+    const char *id = NULL;
     const char *op = NULL;
-    const char *type_s = crm_element_value(expr, PCMK_XA_TYPE);
-    const char *value = crm_element_value(expr, PCMK_XA_VALUE);
-    const char *value_source = crm_element_value(expr, PCMK_XA_VALUE_SOURCE);
+    const char *attr = NULL;
+    const char *type_s = NULL;
+    const char *value = NULL;
+    const char *actual = NULL;
+    const char *source_s = NULL;
+    const char *reference = NULL;
+    char *expanded_attr = NULL;
+    int rc = pcmk_rc_ok;
 
-    enum pcmk__comparison comparison = pcmk__comparison_unknown;
     enum pcmk__type type = pcmk__type_unknown;
+    enum pcmk__reference_source source = pcmk__source_unknown;
+    enum pcmk__comparison comparison = pcmk__comparison_unknown;
+
+    if ((expression == NULL) || (rule_data == NULL)) {
+        return EINVAL;
+    }
+
+    // Get expression ID (for logging)
+    id = pcmk__xe_id(expression);
+    if (pcmk__str_empty(id)) {
```

Comment:
I thought about creating a "valid XML ID" function, but since this is only with schema validation disabled, I figured it's a reasonable relaxation to require ID to simply be non-empty. I could see an argument for being stricter (though even then, I don't think we should bother ensuring IDs are unique across the CIB).

Reply:
I've almost had reason to do that a few times now, and gotten part of the way through implementing it a couple of those. The hardest part is definitely checking for non-ASCII UTF-8 characters that are valid in an ID.